tests/fetch-from-ncbi-virus/filter-and-fields.t

```diff
@@ -0,0 +1,17 @@
+  {"genbank_accession":"X05377","genbank_accession_rev":"X05377.1","database":"GenBank","strain":"","region":"","location":"","collected":"","submitted":"1987-11-29T00:00:00Z","updated":"2016-07-26T00:00:00Z","length":"360","host":"","isolation_source":"","bioproject_accession":"","biosample_accession":"","sra_accession":"","title":"Dengue virus type 2 genomic RNA for NS3 protein N-term","authors":"Biedrzycka,A., Cauchi,M.R., Bartholomeusz,A., Gorman,J.J., Wright,P.J.","submitting_organization":"","publications":"2952760","sequence":"CTCACTGTGTGCTACGTGCTCACTGGACGATCGGCCGATTTGGAACTGGAGAGAGCCGCCGATGTCAAATGGGAAGATCAGGCAGAGATATCAGGAAGCAGTCCAATCCTGTCAATAACAATATCAGAAGATGGTAGCATGTCGATAAAAAACGAAGAGGAAGAACAAACACTGACCATACTCATTAGAACAGGATTGCTGGTGATCTCAGGACTTTTTCCTGTATCAATACCAATCACGGCAGCAGCATGGTACCTGTGGGAAGTGAAGAAACAACGGGCTGGAGTATTGTGGGATGTCCCTTCACCCCCACCCGTGGGAAAGGCTGAACTGGAAGATGGAGCCTATAGAATCAAGCAA","viruslineage_ids":"10239,2559587,2732396,2732406,2732462,2732545,11050,11051,12637,11060"}
+  {"genbank_accession":"X05378","genbank_accession_rev":"X05378.1","database":"GenBank","strain":"","region":"","location":"","collected":"","submitted":"1987-11-29T00:00:00Z","updated":"2016-07-26T00:00:00Z","length":"360","host":"","isolation_source":"","bioproject_accession":"","biosample_accession":"","sra_accession":"","title":"Dengue virus type 2 genomic RNA for NS5 protein N-term","authors":"Biedrzycka,A., Cauchi,M.R., Bartholomeusz,A., Gorman,J.J., Wright,P.J.","submitting_organization":"","publications":"2952760","sequence":"GATCCAATACCCTATGATCCAAAGTTTGAAAAGCAGTTGGGACAAGTAATGCTCCTAGTCCTCTGCGGGACTCAAGTGTTGATGATGAGGACTACATGGGCTCTGTGTGAGGCTTTAACCTTAGCGACCGGGCCTATCTCCACATTGTGGGAAGGAAATCCAGGGAGGTTTTGGAACACTACCATTGCAGTGTCAATGGCTAACATTTTTAGAGGGAGTTACTTGGCCGGAGCTGGACTTCTCTTTTCCATCATGAAGAACACAACCAACACGAGAAGGGGAACTGGCAACATAGGAGAGACGCTTGGAGAGAAATGGAAAAGCCGATTGAACGCATTGGGGAAAAGTGAATTCCAGATC","viruslineage_ids":"10239,2559587,2732396,2732406,2732462,2732545,11050,11051,12637,11060"}
+
+Do the same but without --field.
+This currently fails.
+
+  $ $TESTDIR/../../fetch-from-ncbi-virus 12637 nextstrain/ingest \
+  > --filter='CreateDate_dt:([1987-11-29T00:00:00Z TO 1987-11-29T00:00:01Z])'
+  .* fields\[@\]: unbound variable (re)
+  [1]
```

Comment:
This [fails to match in CI](https://github.com/nextstrain/ingest/actions/runs/6030705991/job/16363034995), but does locally. My guess is this is some sort of difference in syntax support between Bash versions.

It matches locally using both of these versions:

```
% /bin/bash --version
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin21)
Copyright (C) 2007 Free Software Foundation, Inc.

% /opt/homebrew/bin/bash --version
GNU bash, version 5.2.15(1)-release (aarch64-apple-darwin21.6.0)
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
```

I was able to replicate the CI behavior in a GitHub Codespace, which uses:

```
$ /bin/bash --version
GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
```

Reply:
I think it'll be hard to write scripts that support multiple Bash versions, so we should choose a specific version and document that these scripts are only expected to work on that version.

I couldn't figure out an easy way to install Bash 5.0.17 on my Mac. One solution might be to use Docker containers in CI (and locally if needed) to ensure the same version is used everywhere.

Reply:
To clarify for others (because I needed it, but as Victor already knows, I'm sure), the issue is inside `fetch-from-ncbi-virus` and has nothing to do with Cram's pattern matching or the invocation of the command in the Cram test.

In particular, this demonstrates the difference between the two versions:

```
tom@whunk $ bash --version | head -n1
GNU bash, version 4.4.20(1)-release (x86_64-pc-linux-gnu)

tom@whunk $ bash -euo pipefail -c 'declare -a x; printf "<%s>\n" "${x[@]}"'
<>

$ ssh blab-mac

blab@blab-mac ~ % bash --version | head -n1
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin21)

blab@blab-mac ~ % bash -euo pipefail -c 'declare -a x; printf "<%s>\n" "${x[@]}"'
bash: x[@]: unbound variable
```

If we look at the docs for newer Bash (4.4.20):

```
-u      Treat unset variables and parameters other than the spe‐
        cial parameters "@" and "*" as an error when  performing
        parameter  expansion.   If  expansion is attempted on an
        unset variable or parameter, the shell prints  an  error
        message,  and, if not interactive, exits with a non-zero
        status.
```

vs. ancient Bash (3.2.57):

```
-u      Treat unset variables as an error when performing param‐
        eter  expansion.   If expansion is attempted on an unset
        variable, the shell prints an error message, and, if not
        interactive, exits with a non-zero status.
```

It's clear this was an intentional change in Bash.

> I think it'll be hard to write scripts that support multiple Bash versions …

It's a little harder for sure, but doable (and we've done it before). An ancient Bash is basically only an issue in practice for macOS users. So if we don't write for multiple versions, then we're essentially saying you _must_ have upgraded your Bash on macOS one way or another. That might be ok, as it can translate to "we only support the ingest programs inside a Nextstrain runtime (where there is upgraded Bash, even on macOS)". The shebang thing is fixable; we can switch to `#!/usr/bin/env bash` instead and then users can affect which bash they use in a variety of ways.

Reply:
We could certainly also port some of these programs to Python if we wanted… (and deal with Python version issues instead… ;-)

Reply:
This is script that [_has_ to use curl](https://github.com/nextstrain/ncov-ingest/commit/6af81e7561132dcb155dd70ac61a06e491991120)...

If this is an issue, we could remove the script completely and use separate workflow rules. One rule to generate the URL and the next to fetch from the URL stored in an output file.

Reply:
> Unless there's a good reason _not to_ always use `#!/usr/bin/env <x>` I'd suggest we make this our default approach.

Yeah, agreed.

I think we can count on `/usr/bin/env` as much as we can `/bin/bash`. We definitely don't need "the system bash", which is what `/bin/bash` is. We want "the bash the user wants", which will often be the system bash, but not always.

> perhaps a good rule of thumb is that if a script requires newer bash functionality then it's worth writing in Python?

In practice I don't think anything will _require_ newer Bash features. There's usually a fine workaround, usually only slightly more verbose. So I think "worth writing in Python" will largely be decided by other, entirely separate reasons.

> If this is an issue…

It shouldn't be an issue. We should be able to make small adjustments to how `${fields[@]}` is used to be compat across the Bash versions under `set -u`.

Reply:
Thanks for the feedback, all! I've pushed a new set of commits that address the issue here.

I realized that I didn't understand how to use `fetch-from-ncbi-virus` properly in 3fb19172374ce5d86f1737bb9233096ae9a8b7a1 (mentioned in https://github.com/nextstrain/ingest/pull/17#discussion_r1312351367). To clear things up, the issue with the script is: only on ancient Bash, it can't handle cases when any of the options are unspecified. I addressed this in 34089c4c0659dfb55861c3e4a295cd8d5fc9c684 by removing dependence on modern Bash in this specific script.

This is still a good opportunity to address the Bash version problems. 8238f1085f3c446bb40f31c1d96192e4f599b77b...f9ec8311ed45ce79eebc75a858c59705d58e1ca0 provide a clear path for macOS users to use modern Bash if/when needed.

Reply:
> I addressed this in https://github.com/nextstrain/ingest/commit/34089c4c0659dfb55861c3e4a295cd8d5fc9c684 by removing dependence on modern Bash in this specific script.

Awesome!

> This is still a good opportunity to address the Bash version problems. https://github.com/nextstrain/ingest/compare/8238f1085f3c446bb40f31c1d96192e4f599b77b...f9ec8311ed45ce79eebc75a858c59705d58e1ca0 provide a clear path for macOS users to use modern Bash if/when needed.

From that commit range:

> Some scripts may require Bash ≥4. If you are running these scripts on macOS, the builtin Bash (`/bin/bash`) does not meet this requirement.

It seems like we can, and in my opinion should, write our bash scripts such that requiring MacOS users to upgrade bash is not necessary: "There's usually a fine workaround, usually only slightly more verbose". (Recommending them to upgrade is just fine!)

Reply:
> It seems like we can, and in my opinion should, write our bash scripts such that requiring MacOS users to upgrade bash is not necessary: "There's usually a fine workaround, usually only slightly more verbose". (Recommending them to upgrade is just fine!)

This sounds like a good direction going forwards. I don't think every script in this repo was written with that in mind and there's no tests in place to check compatibility, so the snippet in the README is an easy patch to cover potential compatibility issues.